File: packages/mobile-core/lib/core/network/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import '../config/environment.dart';

class ApiResponse<T> {
  final T? data;
  final String? error;

  ApiResponse({this.data, this.error});
}

class ApiClient {
  final String baseUrl;
  final FlutterSecureStorage _storage;
  String? _token;
  bool _initialized = false;

  ApiClient({String? baseUrl, FlutterSecureStorage? storage})
      : baseUrl = baseUrl ?? Environment.apiBaseUrl,
        _storage = storage ?? const FlutterSecureStorage();

  Future<void> init() async {
    if (_initialized) return;
    _token = await _storage.read(key: 'armora_token');
    _initialized = true;
  }

  Future<void> setToken(String? token) async {
    _token = token;
    if (token != null) {
      await _storage.write(key: 'armora_token', value: token);
    } else {
      await _storage.delete(key: 'armora_token');
    }
  }

  Future<String?> getToken() async {
    if (!_initialized) await init();
    return _token;
  }

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (_token != null) 'Authorization': 'Bearer $_token',
      };

  Future<ApiResponse<Map<String, dynamic>>> get(String endpoint) async {
    try {
      final res = await http.get(Uri.parse('$baseUrl$endpoint'),
          headers: _headers);
      if (res.statusCode == 200) {
        return ApiResponse(data: jsonDecode(res.body));
      }
      return ApiResponse(error: 'Error ${res.statusCode}');
    } catch (e) {
      return ApiResponse(error: 'Error de conexion: $e');
    }
  }

  Future<ApiResponse<Map<String, dynamic>>> post(
      String endpoint, Map<String, dynamic> body) async {
    try {
      final res = await http.post(Uri.parse('$baseUrl$endpoint'),
          headers: _headers, body: jsonEncode(body));
      if (res.statusCode == 200 || res.statusCode == 201) {
        return ApiResponse(data: jsonDecode(res.body));
      }
      return ApiResponse(error: 'Error ${res.statusCode}');
    } catch (e) {
      return ApiResponse(error: 'Error de conexion: $e');
    }
  }

  Future<ApiResponse<Map<String, dynamic>>> health() => get('/health');
  Future<ApiResponse<Map<String, dynamic>>> version() => get('/version');
}

File: frontend_web/src/lib/api-client.ts
const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL || 'http://localhost:8885/api/v1';

interface ApiEnvelope<T> {
  data?: T;
  meta?: { requestId: string };
  errors?: Array<{ code: string; message: string; field?: string }>;
}

function hasApiEnvelope<T>(body: unknown): body is ApiEnvelope<T> {
  return Boolean(
    body &&
      typeof body === 'object' &&
      ('data' in body || 'meta' in body || 'errors' in body)
  );
}

function buildHeaders(headers?: HeadersInit) {
  const nextHeaders = new Headers(headers);
  nextHeaders.set('Content-Type', 'application/json');

  if (typeof window !== 'undefined') {
    const token = window.localStorage.getItem('armora_token');
    if (token) {
      nextHeaders.set('Authorization', `Bearer ${token}`);
    }
  }

  return nextHeaders;
}

async function request<T>(endpoint: string, options?: RequestInit): Promise<T> {
  const url = `${API_BASE}${endpoint}`;
  const res = await fetch(url, {
    ...options,
    headers: buildHeaders(options?.headers),
  });

  let body: unknown;
  try {
    body = await res.json();
  } catch {
    if (!res.ok) {
      throw new Error(`Error ${res.status}: el servidor no devolvio JSON`);
    }
    throw new Error('Respuesta inesperada del servidor');
  }

  if (!res.ok) {
    if (hasApiEnvelope<T>(body) && body.errors && body.errors.length > 0) {
      throw new Error(body.errors[0].message);
    }
    throw new Error(`Error ${res.status}`);
  }

  if (hasApiEnvelope<T>(body) && 'data' in body) {
    return body.data as T;
  }

  return body as T;
}

export const api = {
  get: <T>(endpoint: string) => request<T>(endpoint),
  post: <T>(endpoint: string, data: unknown) =>
    request<T>(endpoint, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  put: <T>(endpoint: string, data: unknown) =>
    request<T>(endpoint, {
      method: 'PUT',
      body: JSON.stringify(data),
    }),
  delete: <T>(endpoint: string) =>
    request<T>(endpoint, { method: 'DELETE' }),
  health: () => api.get<{ status: string }>('/health'),
  version: () => api.get<{ version: string }>('/version'),
};
